plugin-server/src/main/ingestion-queues/session-recordings-consumer.ts

```diff
@@ -0,0 +1,202 @@
+import { PluginEvent } from '@posthog/plugin-scaffold'
+import { StatsD } from 'hot-shots'
+import { EachBatchPayload, Kafka } from 'kafkajs'
+
+import { KAFKA_SESSION_RECORDING_EVENTS, KAFKA_SESSION_RECORDING_EVENTS_DLQ } from '../../config/kafka-topics'
+import { PipelineEvent, RawEventMessage } from '../../types'
+import { DependencyUnavailableError } from '../../utils/db/error'
+import { KafkaProducerWrapper } from '../../utils/db/kafka-producer-wrapper'
+import { status } from '../../utils/status'
+import { createPerformanceEvent, createSessionRecordingEvent } from '../../worker/ingestion/process-event'
+import { TeamManager } from '../../worker/ingestion/team-manager'
+import { parseEventTimestamp } from '../../worker/ingestion/timestamps'
+import { instrumentEachBatch, setupEventHandlers } from './kafka-queue'
+
+export const startSessionRecordingEventsConsumer = async ({
+    teamManager,
+    kafka,
+    statsd,
+}: {
+    teamManager: TeamManager
+    kafka: Kafka
+    statsd?: StatsD
+}) => {
+    /*
+        For Session Recordings we need to prepare the data for ClickHouse.
+        Additionally, we process `$performance_event` events which are closely
+        tied to session recording events.
+
+        NOTE: it may be safer to also separate processing of
+        `$performance_event` but for now we'll keep it in the same consumer.
+    */
+
+    // We use our own producer as we want to ensure that we don't have out of
+    // band calls to the `flush` method via the KAFKA_FLUSH_FREQUENCY_MS option.
+    // This ensures that we can handle Kafka Producer errors within the body of
+    // the Kafka consumer handler.
+    const producer = kafka.producer()
+    await producer.connect()
+    const producerWrapper = new KafkaProducerWrapper(producer, statsd, { KAFKA_FLUSH_FREQUENCY_MS: 0 } as any)
+
+    const consumer = kafka.consumer({ groupId: 'session-recordings' })
+    setupEventHandlers(consumer)
+
+    status.info('🔁', 'Starting session recordings consumer')
+
+    await consumer.connect()
+    await consumer.subscribe({ topic: KAFKA_SESSION_RECORDING_EVENTS })
+    await consumer.run({
+        eachBatch: async (payload) => {
+            return await instrumentEachBatch(
+                KAFKA_SESSION_RECORDING_EVENTS,
+                eachBatch({ producer: producerWrapper, teamManager }),
+                payload
+            )
+        },
+    })
+
+    return consumer
+}
+
+export const eachBatch =
+    ({ producer, teamManager }: { producer: KafkaProducerWrapper; teamManager: TeamManager }) =>
+    async ({ batch, heartbeat }: Pick<EachBatchPayload, 'batch' | 'heartbeat'>) => {
+        status.debug('🔁', 'Processing batch', { size: batch.messages.length })
+        for (const message of batch.messages) {
```

Comment:
By using batch it allows us some optimisation e.g. on the `flush` for the producer.

Looking at the logs though it seems that batch sizes are mostly 1 though 🤷  not sure what's happening there.

Re. async I don't think there's much in here that is IO bound. The team id is cached and most of the time we'll already have the team id. We consume from a partition, we produce to a topic. I'm not sure how much gain we'd get and it would add complexity. I suggest we see how it performs as is, and if we need to we can play with e.g. processing partitions in parallel, improve the batching logic on the consumer side as well as on the producer side.

Reply:
If we're not gaining much `eachMessage` would be simpler code, which always makes me happier && later potentially we can make ingestion just be `eachMessage` too